cdap-formats/src/main/java/io/cdap/cdap/format/io/JsonStructuredRecordDatumWriter.java

```diff
@@ -132,61 +137,84 @@ protected void encode(Encoder encoder, Schema schema, Object value) throws IOExc
     Schema nonNullableSchema = schema.isNullable() ? schema.getNonNullable() : schema;
     Schema.LogicalType logicalType = nonNullableSchema.getLogicalType();
 
-    if (value != null && logicalTypeAsString && logicalType != null) {
-      switch (logicalType) {
-        case DATE:
-          Integer date = (Integer) value;
-          // will be encoded to string of format YYYY-mm-DD
-          encoder.writeString(LocalDate.ofEpochDay(date.longValue()).format(DateTimeFormatter.ISO_LOCAL_DATE));
-          break;
-        case TIME_MILLIS:
-          LocalTime localTimeMillis = LocalTime.ofNanoOfDay(TimeUnit.MILLISECONDS.toNanos(((Integer) value)));
-          // will be encoded to string of format HH:mm:ss.SSSSSSSSS
-          encoder.writeString(localTimeMillis.format(DateTimeFormatter.ISO_LOCAL_TIME));
-          break;
-        case TIME_MICROS:
-          LocalTime localTimeMicros = LocalTime.ofNanoOfDay(TimeUnit.MICROSECONDS.toNanos((Long) value));
-          // will be encoded to string of format HH:mm:ss.SSSSSSSSS
-          encoder.writeString(localTimeMicros.format(DateTimeFormatter.ISO_LOCAL_TIME));
-          break;
-        case TIMESTAMP_MILLIS:
-          ZonedDateTime timestampMillis = getZonedDateTime((Long) value, TimeUnit.MILLISECONDS,
-                                                           ZoneId.ofOffset("UTC", ZoneOffset.UTC));
-          // will be encoded to string of format YYYY-mm-DDTHH:mm:ss.SSSSSSSSSZ[UTC]
-          encoder.writeString(timestampMillis.format(DateTimeFormatter.ISO_ZONED_DATE_TIME));
-          break;
-        case TIMESTAMP_MICROS:
-          ZonedDateTime timestampMicros = getZonedDateTime((Long) value, TimeUnit.MICROSECONDS,
-                                                           ZoneId.ofOffset("UTC", ZoneOffset.UTC));
-          // will be encoded to string of format YYYY-mm-DDTHH:mm:ss.SSSSSSSSSZ[UTC]
-          encoder.writeString(timestampMicros.format(DateTimeFormatter.ISO_ZONED_DATE_TIME));
-          break;
-        case DECIMAL:
+    if (value != null && logicalType != null) {
+      // output logical types as string values if needed
+      if (logicalTypeAsString) {
+        handleLogicalTypeAsString(encoder, value, nonNullableSchema, logicalType);
+        return;
+      } else {
+        // encode BigDecimal values using its full numeric representation
+        if (logicalType == Schema.LogicalType.DECIMAL) {
```

Comment:
nit: see if this can be rearranged as series of if-else instead of 3 levels of nested ifs